### PR TITLE
KAFKA-9945: TopicCommand should support --if-exists and --if-not-exists when --bootstrap-server is used

### PR DESCRIPTION
**Pull request by cmccabe**

### CODE REVIEW

**Review by ctan888 — COMMENTED** [review]
I think we should take out ++Set(bootstrapServerOpt) from the last parameter of checkInvalidArgs
```
CommandLineUtils.checkInvalidArgs(parser, options, ifExistsOpt, allTopicLevelOpts -- Set(alterOpt, deleteOpt, describeOpt) ++ Set(bootstrapServerOpt))
CommandLineUtils.checkInvalidArgs(parser, options, ifNotExistsOpt, allTopicLevelOpts -- Set(createOpt) ++ Set(bootstrapServerOpt))
```

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 1]
```diff
@@ -245,7 +245,17 @@ object TopicCommand extends Logging {
 
         newTopic.configs(configsMap)
         val createResult = adminClient.createTopics(Collections.singleton(newTopic))
-        createResult.all().get()
+        try {
+          createResult.all().get()
+        } catch {
+          case e: ExecutionException => {
+            val cause = e.getCause
+            if (cause.isInstanceOf[TopicExistsException] || topic.ifTopicDoesntExist()) {
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/TopicCommand.scala:253]: Shouldn't we re-throw all exceptions except `TopicExistsException` if `topic.ifTopicDoesntExist`?

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 2]
```diff
@@ -245,7 +245,17 @@ object TopicCommand extends Logging {
...
+              throw e
+            }
+          }
+        }
+
         println(s"Created topic ${topic.name}.")
       } else {
         throw new IllegalArgumentException(s"Topic ${topic.name} already exists")
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/TopicCommand.scala:261]: Apparently, we already verify the existence of the topic prior to creating it. Should we also handle `--if-not-exists` in this case?

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 3]
```diff
@@ -257,21 +267,50 @@ object TopicCommand extends Logging {
...
-      val topicsInfo = adminClient.describeTopics(topics.asJavaCollection).values()
-      adminClient.createPartitions(topics.map {topicName =>
-        if (topic.hasReplicaAssignment) {
-          val startPartitionId = topicsInfo.get(topicName).get().partitions().size()
-          val newAssignment = {
-            val replicaMap = topic.replicaAssignment.get.drop(startPartitionId)
-            new util.ArrayList(replicaMap.map(p => p._2.asJava).asJavaCollection).asInstanceOf[util.List[util.List[Integer]]]
+      if(opts.topicConfig.isDefined || opts.configsToDelete.isDefined) {
+        throw new RuntimeException("Using --config or --delete-config is not supported " +
+          "when altering a topic via the broker API.  Use kafka-configs.sh instead.")
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/TopicCommand.scala:272]: nit: Extra space after the first `.`.

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 4]
```diff
@@ -257,21 +267,50 @@ object TopicCommand extends Logging {
...
+      }
+      val tp = new CommandTopicPartition(opts)
+      if (tp.hasPartitions) {
+        println("WARNING: If partitions are increased for a topic that has a key, the partition " +
+          "logic or ordering of the messages will be affected")
+        val topicDescription = try {
+          adminClient.describeTopics(Collections.singleton(tp.name)).
+            all().get().get(tp.name)
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/TopicCommand.scala:280]: nit: unnecessary spaces before `all()`.

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 5]
```diff
@@ -245,7 +245,17 @@ object TopicCommand extends Logging {
 
         newTopic.configs(configsMap)
         val createResult = adminClient.createTopics(Collections.singleton(newTopic))
-        createResult.all().get()
+        try {
+          createResult.all().get()
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/TopicCommand.scala:249]: nit: the parenthesis are not necessary. there is other cases when they can be removed.

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 6]
```diff
@@ -257,21 +267,50 @@ object TopicCommand extends Logging {
...
+              println(s"Ignoring non-existent topic ${tp.name}.")
+              return
+            } else {
+              throw e
+            }
           }
-          topicName -> NewPartitions.increaseTo(topic.partitions.get, newAssignment)
+        }
+        if (topicDescription.partitions().size() == tp.partitions.get) {
+          println(s"Topic ${tp.name} already has ${tp.partitions.get} partitions.  " +
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/TopicCommand.scala:293]: nit: extra space after the `.`. same below.

File: core/src/main/scala/kafka/admin/TopicCommand.scala [hunk 7]
```diff
@@ -639,9 +688,9 @@ object TopicCommand extends Logging {
     private val topicsWithOverridesOpt = parser.accepts("topics-with-overrides",
       "if set when describing topics, only show topics that have overridden configs")
     private val ifExistsOpt = parser.accepts("if-exists",
-      "if set when altering or deleting or describing topics, the action will only execute if the topic exists. Not supported with the --bootstrap-server option.")
+      "if set when altering or deleting or describing topics, the action will only execute if the topic exists.")
```
Comment (dajac) [inline core/src/main/scala/kafka/admin/TopicCommand.scala:691]: We need to update `checkArgs` to accept them with `bootstrap-server`. As the moment, the tool refuses them with the following error:
```
Option "[if-not-exists]" can't be used with option "[bootstrap-server]"
```

**Review by dajac — COMMENTED** [review]
Thanks for the PR. I have left few comments.

**Comment by cmccabe** [comment]
duplicate of #8737

**Pull request closed without merging**